File: Data Structures & Algorithms/duplicate-integer/submission-5.py
class Solution:
    def hasDuplicate(self, nums: List[int]) -> bool:
        if len(nums) == 0:
            return False

        num_count = {}

        for num in nums:
            if num in num_count:
                return True
            else:
                num_count[num] = 1

        return False
            

File: Data Structures & Algorithms/find-target-in-rotated-sorted-array/submission-5.cpp
class Solution {
public:
    int search(vector<int>& nums, int target) {
        int l = 0;
        int r = nums.size() - 1;

        while (l <= r) {
            int m = l + (r - l) / 2;

            if (nums[m] == target) {
                return m;
            };

            // left sorted portion
            if (nums[l] <= nums[m]) {
                if (target < nums[l] || target > nums[m]) {
                    l = m + 1;
                } else {
                    r = m - 1;
                };
            } else { // right sorted portion
                if (target > nums[r] || target < nums[m]) {
                    r = m - 1;
                } else {
                    l = m + 1;
                };
            };
        };

    return -1;
    
    };
};
